### PR TITLE
add new sequence_bias param type for SequenceBiasLogitsProcessor

### PR DESCRIPTION
**Pull request by VladOS95-cyber**
# What does this PR do?
Introduce new sequence_bias parameter type as List[List[List[int], float]] for SequenceBiasLogitsProcessor

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@gante 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by VladOS95-cyber** [comment]
Hi @gante, please, take a look at my PR.

File: docs/source/en/chat_templating.md [hunk 1]
```diff
@@ -196,7 +196,7 @@ Not all models require generation prompts. Some models, like LLaMA, don't have a
 special tokens before bot responses. In these cases, the `add_generation_prompt` argument will have no effect. The exact
 effect that `add_generation_prompt` has will depend on the template being used.
 
-## What does "continue_last_message" do?
+## What does "continue_final_message" do?
```
Comment (gante) [inline docs/source/en/chat_templating.md:199]: The changes in this file are unrelated to the PR. If you intentionally added these lines, please open a separate PR 🤗 If not, rebasing may help

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1064,8 +1064,8 @@ class SequenceBiasLogitsProcessor(LogitsProcessor):
     </Tip>
 
     Args:
-        sequence_bias (`Dict[Tuple[int], float]`):
-            Dictionary that maps a sequence of tokens to its bias term. Positive biases increase the odds of the
+        sequence_bias (`Union[Dict[Tuple[int], float], List[List[Union[List[int], float]]]]`):
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1067]:
```suggestion
        sequence_bias (`List[List[Union[List[int], float]]]`):
```
let's not document the case that doesn't support serialization, even though we can accept the format :) that way, we reduce the odds of users reaching the serialization issue

Also related to this suggestion: there should be no trace of the previous format in the docstring (including in the example below) and type hints

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -1178,11 +1179,15 @@ def _prepare_bias_variables(self, scores: torch.FloatTensor):
 
     def _validate_arguments(self):
         sequence_bias = self.sequence_bias
-        if not isinstance(sequence_bias, dict) or len(sequence_bias) == 0:
-            raise ValueError(f"`sequence_bias` has to be a non-empty dictionary, but is {sequence_bias}.")
-        if any(not isinstance(sequence_ids, tuple) for sequence_ids in sequence_bias.keys()):
+        if not isinstance(sequence_bias, dict) and not isinstance(sequence_bias, list) or len(sequence_bias) == 0:
+            raise ValueError(
+                f"`sequence_bias` has to be a non-empty dictionary, , or non-empty list of lists but is {sequence_bias}."
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1184]:
```suggestion
                f"`sequence_bias` has to be a non-empty dictionary or non-empty list of lists, but is {sequence_bias}."
```

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1191,9 +1196,30 @@ def _validate_arguments(self):
...
+                not (
+                    isinstance(token_bias_pair, list)
+                    and all(isinstance(token_id, int) and token_id > 0 for token_id in token_bias_pair)
+                    or isinstance(token_bias_pair, float)
+                )
+                for token_bias_pair in sequence
+            )
+            or len(sequence) == 0
+            for sequence in sequence_bias
+        ):
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1210]: This one-liner `if` is hard to read, can we break it into multiple variables?

e.g. `all_token_bias_pairs_are_valid = ...`

(readability helps with long-term maintenance 🤗 )

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -1191,9 +1196,30 @@ def _validate_arguments(self):
...
+            raise ValueError(
+                f"Each element in `sequence_bias` has to be a non-empty list of lists of positive integers and float, but is "
+                f"{sequence_bias}."
+            )
+        if isinstance(sequence_bias, dict) and any(not isinstance(bias, float) for bias in sequence_bias.values()):
             raise ValueError(f"`sequence_bias` has to be a dict with floats as values, but is {sequence_bias}.")
 
+    def _convert_list_arguments_into_dict(self):
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1218]:
```suggestion
    def _convert_list_arguments_into_dict(self):
        """BC: we used to accept `dict{tuple of tokens: float}`"""
```
This will help us track the effort to keep backwards compatibility

File: tests/generation/test_configuration_utils.py [hunk 6]
```diff
@@ -224,6 +224,18 @@ def test_generation_mode(self):
         config = GenerationConfig()
         self.assertEqual(config.get_generation_mode(assistant_model="foo"), GenerationMode.ASSISTED_GENERATION)
 
+    def test_serialize_generation_sequence_bias(self):
```
Comment (gante) [inline tests/generation/test_configuration_utils.py:227]: Request: add a one-line test docstring explaining what the test does

(we're making sure new tests are briefly explained)

**Review by gante — COMMENTED** [review]
Hi @VladOS95-cyber 👋 Thank you for taking care of the issue and opening the PR!

The PR looks mostly good to me -- I've added a few comments to help us with long-term maintainability 🤗

File: src/transformers/generation/logits_process.py (hunk 2 above)
Comment (VladOS95-cyber) [reply src/transformers/generation/logits_process.py:1067]: ok, done
